cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/AbstractInMemoryProgramRunner.java

```diff
@@ -84,7 +89,8 @@ protected final ProgramController startAll(Program program, ProgramOptions optio
         components.put(program.getName(), instanceId, controller);
       }
 
-      return new InMemoryProgramController(components, program, options, runId);
+      ProgramStateWriter programStateWriter = new DirectStoreProgramStateWriter(runtimeStore);
```

Comment:
Shouldn't be creating this everytime. Make the class `DirectStoreProgramStateWriter` guice injectable and have the constructor takes `ProgramStateWriter` interface and setup the right binding in the app fabric module and (and the modules used in twill container, which binds to a different implementation).